akka-sample-kafka-to-sharding/README.md

```diff
@@ -0,0 +1,191 @@
+[info] [2020-01-17 08:27:58,199] [INFO] [akka://KafkaToSharding/user/kafka-event-processor] - Forwarding message for entity 29 to cluster sharding
+[info] [2020-01-17 08:27:58,204] [INFO] [akka://KafkaToSharding/system/sharding/user-processing/45/29] - user 29 purchase cat t-shirt, quantity 1, price 2093
+[info] [2020-01-17 08:28:08,218] [INFO] [akka://KafkaToSharding/user/kafka-event-processor] - Forwarding message for entity 56 to cluster sharding
+[info] [2020-01-17 08:28:08,218] [INFO] [akka://KafkaToSharding/system/sharding/user-processing/6/56] - user 56 purchase akka t-shirt, quantity 3, price 8576
+[info] [2020-01-17 08:28:28,288] [INFO] [akka://KafkaToSharding/user/kafka-event-processor] - Forwarding message for entity 44 to cluster sharding
+[info] [2020-01-17 08:28:28,296] [INFO] [akka://KafkaToSharding/system/sharding/user-processing/59/44] - user 44 purchase cat t-shirt, quantity 3, price 9716
+```
+
+Each forwarding messaging is followed by log for the same entity on the current node.
+
+Using Akka management we can see the shard allocations and the number of entities per shard (uses `curl` and `jq`):
+
+```
+
+// Node 1:
+ curl -v localhost:8551/cluster/shards/user-processing  | jq
+
+// Node 2:
+ curl -v localhost:8552/cluster/shards/user-processing  | jq
+```
+
+We can count the number of shards on each:
+
+```
+curl -v localhost:8551/cluster/shards/user-processing  | jq -r "." | grep shardId  | wc
+```
+
+The number of shards will depend on which entities have received messages.
+
+We now have a 2 node Akka Cluster with a Kafka Consumer running on each where the kafka partitions align 
+with Cluster shards.
+
+A use case for sending the processing to sharding is it allows each entity to be queried from any where in the cluster
+e.g. from a HTTP or gRPC front end.
+
+The sample includes a gRPC front end that gets the running total of number of purchases and total money spent
+by each customer. Requests can come via gRPC on any node for any entity but sharding will route them to
+the correct node even if that moves due to a kafka rebalance.
+
+A gRPC client is included which can be started with...
+
+```
+ sbt "client/run"
+```
+
+It assumes there is one of the nodes running its front end port on 8081. 
+You can enter the user id to get their stats: The users are `0-199`, try entering users
+that have shown up in the logs of the other processes.
+
+```
+7
+User 7 has made 2 for a total of 3096p
+Enter user id or :q to quit
+3
+User 3 has made 1 for a total of 12060p
+Enter user id or :q to quit
+4
+User 4 has made 1 for a total of 7876p
+Enter user id or :q to quit
+5
+User 5 has made 0 for a total of 0p
+Enter user id or :q to quit
+1
+User 1 has made 0 for a total of 0p
+Enter user id or :q to quit
+```
+
+We've now demonstrated two things:
+
+* Keeping the processing local, where ever the Kafka partition is consumed the shard will be moved to that location
+* The state for each entity is globally accessible from all nodes 
```

Comment:
I was able to query entities that were local to the node I connected to, but when I queried entities on the second processor node I received a fatal exception.  I'm not that familiar with akka gRPC, but from the looks of `UserGrpcService` I would expect it to be able to use the shard coordinator to make calls over cluster.

```
[info] running client.ClientApp
Enter user id or :q to quit
31
User 31 has made 1 purchases for a total of 9500p
Enter user id or :q to quit
45
User 45 has made 2 purchases for a total of 1448p
Enter user id or :q to quit
45
User 45 has made 2 purchases for a total of 1448p
Enter user id or :q to quit
22
[error] (run-main-0) io.grpc.StatusRuntimeException: INTERNAL
[error] io.grpc.StatusRuntimeException: INTERNAL
[error]         at io.grpc.Status.asRuntimeException(Status.java:533)
[error]         at akka.grpc.internal.UnaryCallAdapter.onClose(UnaryCallAdapter.scala:40)
[error]         at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
[error]         at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
[error]         at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
[error]         at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:700)
[error]         at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
[error]         at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
[error]         at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
[error]         at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:399)
[error]         at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:510)
[error]         at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:66)
[error]         at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:630)
[error]         at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$700(ClientCallImpl.java:518)
[error]         at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:692)
[error]         at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:681)
[error]         at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
[error]         at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
[error]         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error]         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error]         at java.lang.Thread.run(Thread.java:748)
[error] stack trace is suppressed; run last client / Compile / bgRun for the full output
```

Reply:
Let me see what's going on. There should have been some logs on the node you connected to